packages/@atjson/document/src/index.ts

```diff
@@ -24,6 +26,43 @@ export default class AtJSON {
     this.contentType = options.contentType || 'text/plain';
     this.queries = [];
     this.schema = options.schema || {};
+
+    this.changeListeners = [];
+  }
+
+  /**
+   * I'm on a plane; I'm not sure the best approach to cross-platform event
+   * listeners and don't have internet access at the moment, so I'm just going
+   * to quickly roll my own here. To be updated.
+   */
+  addEventListener(eventName: string, func: Function): void {
+    if (eventName !== 'change') throw new Error('Unsupported event. `change` is the only constant.');
+    this.changeListeners.push(func);
+  }
+
+  /*
+  removeEventListener(eventName: string, func: Function): void {
+    throw new Error('Unimplemented.');
+  }
+  */
+
+  /**
+   * This is really coarse, just enough to allow different code in the editor to detect
+   * changes in the document without handling that change management separately.
+   *
+   * Eventually it should be possible to handle this transactionally, but for
+   * now we batch all changes enacted within one cycle of the event loop and
+   * fire the change event only once. n.b that we don't send any information
+   * about the changes here yet, but that's not to say we couldn't, but rather
+   * it's not clear right now what the best approach would be so it's left
+   * undefined.
+   */
+  private triggerChange() {
+    if (this.pendingChangeEvent) return;
+    this.pendingChangeEvent = setTimeout(_ => {
+      this.changeListeners.forEach(l => l());
+      delete this.pendingChangeEvent;
+    }, 0)
```

Comment:
Yeah - it shouldn't be possible to interrupt the document transform from listeners. We may want to explicitly catch within the forEach loop (not sure if throwing in one would cause issues for others?)